expr/func_aspercent_test.go

```diff
@@ -0,0 +1,649 @@
+		},
+		t,
+		math.NaN(),
+		nil,
+		nil,
+		"None",
+	)
+}
+
+func TestAsPercentTotalFloat(t *testing.T) {
+	total := 123.456
+	out := []schema.Point{
+		{Val: 0, Ts: 10},
+		{Val: 0, Ts: 20},
+		{Val: 5.5 / total * 100, Ts: 30},
+		{Val: math.NaN(), Ts: 40},
+		{Val: math.NaN(), Ts: 50},
+		{Val: 1234567890 / total * 100, Ts: 60},
+	}
+	testAsPercent(
+		"single-noarg",
+		[]models.Series{
+			{
+				Interval:   10,
+				QueryPatt:  "func(a;tag=something;tag2=anything)",
+				Target:     "a;tag=something;tag2=anything",
+				Datapoints: getCopy(a),
+			},
+		},
+		[]models.Series{
+			{
+				Interval:   10,
+				QueryPatt:  "asPercent(func(a;tag=something;tag2=anything),123.456)",
+				Target:     "asPercent(a;tag=something;tag2=anything,123.456)",
+				Datapoints: out,
+			},
+		},
+		t,
+		total,
+		nil,
+		nil,
+		"123.456",
+	)
+}
+
+func TestAsPercentTotalSerie(t *testing.T) {
+	out1 := []schema.Point{
+		{Val: math.NaN(), Ts: 10},
+		{Val: math.NaN(), Ts: 20},
+		{Val: math.Inf(0), Ts: 30},
+		{Val: math.NaN(), Ts: 40},
+		{Val: math.NaN(), Ts: 50},
+		{Val: math.NaN(), Ts: 60},
+	}
+	out2 := []schema.Point{
+		{Val: math.NaN(), Ts: 10},
+		{Val: math.NaN(), Ts: 20},
+		{Val: 199 / 5.5 * 100, Ts: 30},
+		{Val: math.NaN(), Ts: 40},
+		{Val: math.NaN(), Ts: 50},
+		{Val: float64(250) / 1234567890 * 100, Ts: 60},
+	}
+	testAsPercent(
+		"multi-serietotal",
+		[]models.Series{
+			{
+				Interval:   10,
+				QueryPatt:  "func(tag=something;tag2=anything)",
+				Target:     "b;tag=something;tag2=anything",
+				Datapoints: getCopy(b),
+			},
+			{
+				Interval:   10,
+				QueryPatt:  "func(tag=something;tag2=anything)",
+				Target:     "d;tag=something;tag2=anything",
+				Datapoints: getCopy(d),
+			},
+		},
+		[]models.Series{
+			{
+				Interval:   10,
+				QueryPatt:  "asPercent(func(tag=something;tag2=anything),func(tag=some;tag2=totalSerie))",
+				Target:     "asPercent(b;tag=something;tag2=anything,a;tag=some;tag2=totalSerie)",
+				Datapoints: out1,
+			},
+			{
+				Interval:   10,
+				QueryPatt:  "asPercent(func(tag=something;tag2=anything),func(tag=some;tag2=totalSerie))",
+				Target:     "asPercent(d;tag=something;tag2=anything,a;tag=some;tag2=totalSerie)",
+				Datapoints: out2,
+			},
+		},
+		t,
+		math.NaN(),
+		[]models.Series{
+			{
+				Interval:   10,
+				QueryPatt:  "func(tag=some;tag2=totalSerie)",
+				Target:     "a;tag=some;tag2=totalSerie",
+				Datapoints: getCopy(a),
+			},
+		},
+		nil,
+		"serie",
+	)
+}
+
+func TestAsPercentTotalSeries(t *testing.T) {
+	out1 := []schema.Point{
+		{Val: math.NaN(), Ts: 10},
+		{Val: math.NaN(), Ts: 20},
+		{Val: math.Inf(0), Ts: 30},
+		{Val: math.NaN(), Ts: 40},
+		{Val: math.NaN(), Ts: 50},
+		{Val: math.NaN(), Ts: 60},
+	}
+	out2 := []schema.Point{
+		{Val: math.NaN(), Ts: 10},
+		{Val: math.NaN(), Ts: 20},
+		{Val: float64(199) * 100, Ts: 30},
+		{Val: float64(29) / 2 * 100, Ts: 40},
+		{Val: float64(80) / 3.0 * 100, Ts: 50},
+		{Val: float64(250) / 4 * 100, Ts: 60},
```

Comment:
shouldn't out2 be the same as in the previous function? because both cases have `asPercent(d,a)`

Reply:
No, because when totals is a seriesList with `len(totals) == len(series)`, the series first get sorted by tag before getting matched. So, in this test case it would be asPercent(d,c) and asPercent(b,a). This behavior is same in Graphite